python/sources/simple-csv-new/connector/source_connector.py

```diff
@@ -0,0 +1,178 @@
+        value_parser: Optional[Callable[[dict], MessageValue]] = json.dumps,
+        **csv_reader_kwargs,
+    ):
+        """
+        :param path: file path to a .csv file
+        :param key_parser: how a key is generated for a given data row. Options:
+            - None; no key will be generated (default)
+            - A string, representing a column name; uses that row's column value.
+            - A function that handles the row as a dict.
+        :param value_parser: A function for generating resulting message value for the
+            row. Default: json.dumps
+        :param csv_reader_kwargs: kwargs passthrough to csv.DictReader
+        """
+        self._path = path
+        self._key_parser = key_parser
+        self._value_parser = value_parser
+        self._file = None
+        self._reader = None
+        self._reader_kwargs = csv_reader_kwargs
+
+    def _as_message(self, data: dict) -> KafkaMessage:
+        """Convenience method to create a kafka-formatted message"""
+        return KafkaMessage(
+            key=self.key_parser(data),
+            value=self.value_parser(data),
+            headers=None
+        )
+
+    def _consume(self) -> Iterator[dict]:
+        """Lazy row reading of the CSV to avoid loading it all in memory"""
+        for row in self._reader:
+            yield row
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.stop()
+
+    def start(self):
+        if not self._file:
+            self._file = open(self._path, "r")
+            self._reader = csv.DictReader(self._file, **self._reader_kwargs)
+            logger.info(f"CSV Consumer beginning reading file: {self._path}")
+
+    def stop(self):
+        if self._file:
+            logger.info(f"CSV Consumer has stopped reading file: {self._path}")
+            self._file.close()
+            self._file = None
+        self._reader = None
+
+    def key_parser(self, data: dict) -> Optional[MessageKey]:
+        """
+        :param data: CSV row data as a dict
+
+        :return: a MessageKey type or None
+        """
+        if self._key_parser:
+            if isinstance(self._key_parser, str):
+                return data[self._key_parser]
+            return self._key_parser(data)
+
+    def value_parser(self, data: dict) -> Optional[MessageValue]:
+        """
+        :param data: CSV row data as a dict
+
+        :return: a MessageValue type or None
+        """
+        return self._value_parser(data)
+
+    def consume(self) -> Optional[KafkaMessage]:
+        """
+        Get a CSV row as a producer-ready message
+
+        :return: a KafkaMessage type when row is available, else None
+        """
+        try:
+            return self._as_message(next(self._consume()))
+        except StopIteration:
+            logger.info("CSV Consumer has reached the end of the file")
+            return
+
+
+class CsvSourceConnector(SourceConnector):
```

Comment:
I struggle to see for now the value of this class. It adds massive amount of boilerplate to the code. Can we merge it with SourceConsumer or at least pass directly producer and consumer so we don't have that ugly __init__?

Reply:
Daniil agrees, will change this to something else